src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -303,7 +305,7 @@ public CompilationAnalysisContext(Compilation compilation, AnalyzerOptions optio
         /// <param name="diagnostic"><see cref="Diagnostic"/> to be reported.</param>
         public void ReportDiagnostic(Diagnostic diagnostic)
         {
-            DiagnosticAnalysisContextHelpers.VerifyArguments(diagnostic);
+            DiagnosticAnalysisContextHelpers.VerifyArguments(diagnostic, _isSupportedDiagnostic);
             lock (_reportDiagnostic)
```

Comment:
Why do we lock on a delegate object passed to the constructor from outside? It introduces an implicit dependency on the calling code (e.g. it might pass the same object to other unrelated methods that can also lock on it).